File: default/server/_jscript/RequestContext.js
(function() {
  var node_path, node_uuid, smio;
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };
  node_path = require('path');
  node_uuid = require('node-uuid');
  smio = global.smoothio;
  smio.RequestContext = (function() {
    function RequestContext(server, uri, httpRequest, httpResponse, adminDB, sharedDB, serverDB) {
      this.server = server;
      this.uri = uri;
      this.httpRequest = httpRequest;
      this.httpResponse = httpResponse;
      this.adminDB = adminDB;
      this.sharedDB = sharedDB;
      this.serverDB = serverDB;
      this.inst = this.server.inst;
      this.cookies = smio.Util.Server.parseCookies(this.httpRequest.headers['cookie']);
    }
    RequestContext.prototype.handleRequest = function() {
      var cfgKey, cfgVal, ctype, fname, hasHandler, respHeaders;
      this.inst.lastRequestTime = new Date;
      respHeaders = {};
      try {
        if (hasHandler = this.uri.pathItems.length && this.uri.pathItems[0] === '_' && this.uri.pathItems.length >= 2) {
          switch (this.uri.pathItems[1]) {
            case "dynfile":
              if ((cfgKey = this.uri.query['config'])) {
                if (cfgKey === '_res.js') {
                  respHeaders['Content-Type'] = 'text/javascript';
                  this.serveFile('_scripts/_res.js', respHeaders);
                } else if ((cfgVal = '' + smio.Util.Object.select(this.server.inst.config, cfgKey)) && (fname = this.uri.query[cfgVal])) {
                  if ((ctype = this.uri.query['type'])) {
                    respHeaders['Content-Type'] = ctype;
                  }
                  this.serveFile(fname, respHeaders);
                }
              } else {
                respHeaders['Content-Type'] = 'text/plain';
                this.httpResponse.writeHead(404, respHeaders);
                this.httpResponse.end("404 File Not Found: " + (node_path.join(this.server.fileServer.root, fname)) + " (dynamic file)");
              }
              break;
            case "file":
              this.serveFile(this.uri.pathItems.slice(2).join('/'), respHeaders);
              break;
            default:
              hasHandler = false;
          }
        }
        if (!hasHandler) {
          return this.serveFile('smoothio.html', respHeaders);
        }
      } catch (err) {
        respHeaders['Content-Type'] = 'text/plain';
        this.httpResponse.writeHead(500, respHeaders);
        return this.httpResponse.end("500 Internal Server Error:\n" + (this.inst.formatError(err)));
      }
    };
    RequestContext.prototype.serveFile = function(filePath, respHeaders) {
      if (node_path.existsSync(node_path.join(this.server.fileServer.root, filePath))) {
        return this.server.fileServer.serveFile(filePath, 200, respHeaders, this.httpRequest, this.httpResponse).addListener('error', __bind(function(err) {
          respHeaders['Content-Type'] = 'text/plain';
          this.httpResponse.writeHead(err.status, smio.Util.Object.mergeDefaults(err.headers, respHeaders));
          return this.httpResponse.end(JSON.stringify(err));
        }, this));
      } else {
        respHeaders['Content-Type'] = 'text/plain';
        this.httpResponse.writeHead(404, respHeaders);
        return this.httpResponse.end("404 File Not Found: " + (node_path.join(this.server.fileServer.root, filePath)));
      }
    };
    return RequestContext;
  })();
}).call(this);

File: _cscript/RequestContext.cs

require './Session'
_ = require 'underscore'
node_path = require 'path'
node_uuid = require 'node-uuid'
node_urlq = require 'querystring'
smio = global.smoothio

class smio.RequestContext
	@parseSmioCookie: (cookies, fail) ->
		if _.isString cookies
			cookies = smio.Util.Server.parseCookies cookies
		parse = -> JSON.parse node_urlq.unescape cookies['smoothio']
		if fail
			smioCookie = parse()
		else
			try
				smioCookie = parse()
			catch err
				smioCookie = {}
		smioCookie

	constructor: (@server, @uri, @httpRequest, @httpResponse, @adminDB, @sharedDB, @serverDB) ->
		@inst = @server.inst
		@postData = null
		@smioCookie = smio.RequestContext.parseSmioCookie @cookies = smio.Util.Server.parseCookies @httpRequest.headers['cookie']
		if @httpRequest.method is 'POST'
			@httpRequest.on 'end', () =>
				@handleRequest()
			@httpRequest.on 'data', (data) =>
				@postData += "#{data}"
		else
			@handleRequest()

	handleRequest: ->
		#date.setTime date.getTime() + (@server.inst.config.session.timeout * 60 * 1000)
		#expires=#{smio.Util.DateTime.addMinutes(smio.Util.Number.tryParseInt(@inst.config.session.timeout, 20)).toGMTString()}
		@inst.lastRequestTime = new Date
		if not @smioCookie['sessid']
			@smioCookie['sessid'] = node_uuid()
		respHeaders =
			'Set-Cookie': "smoo=#{node_urlq.escape JSON.stringify @smioCookie}; path=/"
		try
			if hasHandler = @uri.pathItems.length and @uri.pathItems[0] is '_' and @uri.pathItems.length >= 2
				switch @uri.pathItems[1]
					when "poll"
						respHeaders['Content-Type'] = 'text/plain'
						finish = (data) =>
							@httpResponse.writeHead 200, respHeaders
							@httpResponse.end JSON.stringify data
						if @uri.pathItems[2] is 'f'
							(smio.Session.getBySessionID @server, @smioCookie['sessid']).handleFetch @, null, finish
						else
							finish {}
					when "dynfile"
						if (cfgKey = @uri.query['config'])
							if cfgKey is '_res.js'
								respHeaders['Content-Type'] = 'text/javascript'
								@serveFile '_scripts/_res.js', respHeaders
							else if (cfgVal = '' + smio.Util.Object.select @server.inst.config, cfgKey) and (fname = @uri.query[cfgVal])
								if (ctype = @uri.query['type'])
									respHeaders['Content-Type'] = ctype
								@serveFile fname, respHeaders
						else
							respHeaders['Content-Type'] = 'text/plain'
							@httpResponse.writeHead 404, respHeaders
							@httpResponse.end "404 File Not Found: #{node_path.join @server.fileServer.root, fname} (dynamic file)"
					when "file"
						@serveFile (@uri.pathItems[2...].join '/'), respHeaders
					else
						hasHandler = false
			if not hasHandler
				@serveFile 'smoothio.html', respHeaders
		catch err
			respHeaders['Content-Type'] = 'text/plain'
			@httpResponse.writeHead 500, respHeaders
			@httpResponse.end "500 Internal Server Error:\n#{@inst.formatError err}"

	serveFile: (filePath, respHeaders) ->
		if node_path.existsSync node_path.join @server.fileServer.root, filePath
			@server.fileServer.serveFile(filePath, 200, respHeaders, @httpRequest, @httpResponse).addListener 'error', (err) =>
				respHeaders['Content-Type'] = 'text/plain'
				@httpResponse.writeHead err.status, smio.Util.Object.mergeDefaults err.headers, respHeaders
				@httpResponse.end JSON.stringify err
		else
			respHeaders['Content-Type'] = 'text/plain'
			@httpResponse.writeHead 404, respHeaders
			@httpResponse.end "404 File Not Found: #{node_path.join @server.fileServer.root, filePath}"

